lib/gauge_polyakov_loop.cu

```diff
@@ -0,0 +1,276 @@
+#include <gauge_field.h>
+#include <gauge_tools.h>
+#include <instantiate.h>
+#include <tunable_nd.h>
+#include <tunable_reduction.h>
+#include <kernels/gauge_polyakov_loop.cuh>
+
+namespace quda {
+
+  template<typename Float, int nColor, QudaReconstructType recon>
+  class GaugeInsertTimeslice : public TunableKernel2D {
+    GaugeField &u;
+    const GaugeField &s;
+    int timeslice;
+    unsigned int minThreads() const { return s.LocalVolumeCB(); }
+
+  public:
+    GaugeInsertTimeslice(GaugeField &u, const GaugeField &s, int timeslice) :
+      TunableKernel2D(u, 2),
+      u(u),
+      s(s),
+      timeslice(timeslice)
+    {
+      strcat(aux, ",4d_vol=");
+      strcat(aux, u.VolString());
+
+      apply(device::get_default_stream());
+    }
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      GaugeInsertTimesliceArg<Float, nColor, recon> arg(u, s, timeslice);
+      launch<InsertTimeslice>(tp, stream, arg);
+    }
+
+    long long flops() const
+    {
+      // just a copy
+      return 0;
+    }
+
+    long long bytes() const {
+      // load timeslice, store into u
+      return 2ll * s.Bytes();
+    }
+  };
+
+  template<typename Float, int nColor, QudaReconstructType recon>
+  class GaugePolyakovLoopProduct : public TunableKernel2D {
+    GaugeField &product_field;
+    const GaugeField &u;
+    unsigned int minThreads() const { return product_field.LocalVolumeCB(); }
+
+  public:
+    GaugePolyakovLoopProduct(const GaugeField &u, GaugeField &product_field) :
+      TunableKernel2D(u, 2),
+      product_field(product_field),
+      u(u)
+    {
+      strcat(aux, ",4d_vol=");
+      strcat(aux, u.VolString());
```

Comment:
can delete the 4-d volume part of this since that will already be in the `TuneKey`

Reply:
Same as above, but this one's less critical since the relevant info can be extracted from the 3-d volume. Done locally.